server/src/main/java/org/elasticsearch/action/support/UnsafePlainActionFuture.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.support;
+
+import org.elasticsearch.common.util.concurrent.EsExecutors;
+import org.elasticsearch.core.CheckedConsumer;
+
+import java.util.Objects;
+
+/**
+ * An unsafe future. You should not need to use this for new code, rather you should be able to convert that code to be async
+ * or use a clear hierarchy of thread pool executors around the future.
+ *
+ * This future is unsafe, since it allows notifying the future on the same thread pool executor that it is being waited on. This
+ * is a common deadlock scenario, since all threads may be waiting and thus no thread may be able to complete the future.
+ */
+public class UnsafePlainActionFuture<T> extends PlainActionFuture<T> {
```

Comment:
I opted to add this unsafe variant that is then used in all places where I found conflicts, since fixing them all would make this PR size unmanageable. Using this we can now fix them one by one, spreading out the load too.

Reply:
:+1: can we mark this as `@Deprecated(forRemoval = true)` so that IDEs highlight its usages?

Reply:
Do we need a task/issue/ticket on fixing them one-by-one?

Reply:
I think one by one is too granular. I'll probably create area level ones (perhaps a bit more granular, for instance the one in `AbstractClient` needs a separate one).